### ISSUE TITLE
mrview - segmentation fault with -overlay.load

### ISSUE THREAD

**Issue by ansk** [MRtrix3/mrtrix3#871]
Dear experts,

I encountered segmentation fault when running mrview with option -overlay.load.

The behaviour is very strange, the error occurs in only some precise conditions:
1. when main image is 4D dataset AND when overlay image is in .nii.gz format, or
2. when both main image and overlay image are 4D datasets.

AND

commandline is NOT using -load option, i.e. command is like:
mrview main_image -overlay.load overlay_image

In other cases when main image is 3D dataset (both main image and overlay image is 3D; main image is 3D, overlay image is 4D) it seems to work OK.

Loading the images with overlay tool in GUI or with specifying -load option works OK in all tested cases (i.e. also with images where loading via -overlay.load produces segmentation fault).

Also, when -load option is NOT specified, i.e. command is like:
mrview main_mage -nointerpolate
the -nointeropolate option is ignored, which is quite confusing.

My version of mrtrix is following:
0.3.15-370-gff066551

Regards,

Antonin Skoch

**Comment by Lestropie** [comment]
I wasn't able to reproduce this issue on any branch on my Windows machine, Eigen 3.2 or 3.3.

I've also been unable to find the commit corresponding to your specific version; are you currently checked out on some other branch? Can you try reproducing the fault with an up-to-date installation?

Regarding `mrview`'s `-nointerpolate` option, it looks like this has already been fixed. I haven't pinpointed exactly where since the code has changed a bit: the option is now `-interpolation` and it takes a boolean value as input.

**Comment by ansk** [comment]
I have updated to:
MRtrix 0.3.15-457-g80f4abf0
and the problem with segmentation fault is still persisting.

I also tried to run:
`mrview image.nii.gz -interpolation 0` (i.e. without specifying `-load` ) and the `-interpolation 0` flag is ignored, image gets interpolated. 

When specifying `-load`, i.e. with command line
`mrview -load image.nii.gz -interpolation 0`
the interpolation is correctly switched off.

Attached find my configure.log and build.log:

[build_configure.tar.gz](https://github.com/MRtrix3/mrtrix3/files/681905/build_configure.tar.gz)

**Comment by maxpietsch** [comment]
I could not reproduce the interpolation problem with mrview 0.3.15-457-g80f4abf0: `mrview 1.nii.gz -interpolation 0` switches interpolation off on my machine.

Please note that `-interpolation 0` is applied to only the active image. `mrview 1.nii.gz -interp 0 -load 1.mif` will cause `1.nii.gz` not to be interpolated and `1.mif` to be interpolated. The opposite is the case if you use `mrview 1.nii.gz -load 1.mif  -interp 0`.

**Comment by ansk** [comment]
@maxpietsch , does your 1.nii.gz is 3D or 4D?
I recently found that the `-interpolation 0` does not work only when the image is 4D and `.nii.gz`. In case of 3D images or in case of 4D `.mif` images it works OK. It seems that this problem is related with above-mentioned segmentation fault problem where I observer similar behaviour.

**Comment by maxpietsch** [comment]
I used a 3D volume and, to make sure there are not race conditions, I also used a 4D `nii.gz` and `mif` image with 1350 volumes (2.8GB uncompressed).

**Comment by maxpietsch** [comment]
I don't know what is going on but if you have an image where the segfault happens you could try to get to the bottom of it with:

    ./configure -assert assert
    ./build assert/bin/mrview
    assert/bin/mrview problemimage.nii.gz -overlay.load another_problem_image.nii.gz

**Comment by ansk** [comment]
`-interpolation 0` does not work with my installation also with `.mif.gz` images. Both in `.nii` and `.mif` works OK.

Regarding segmentation fault problem, I tried to compile `mrview` with assert and run like you suggested. It did not produce any specific output, only segmentation fault. 
Running 
`assert/bin/mrview -debug image1.nii.gz -overlay.load image2.nii.gz`
produced following output before segmentation fault:

```
mrview: [DEBUG] No config file found at "/etc/mrtrix.conf"
mrview: [DEBUG] No config file found at ".mrtrix.conf"
mrview: [INFO] GL renderer:  Quadro FX GRID K1/PCIe/SSE2
mrview: [INFO] GL version:   3.3.0 NVIDIA 352.39
mrview: [INFO] GL vendor:    NVIDIA Corporation
mrview: [DEBUG] loading font into OpenGL texture...
mrview: [DEBUG] compiling OpenGL vertex shader:
#version 330 core
layout(location = 0) in vec2 pos;
layout(location = 1) in vec2 font_pos;
uniform float scale_x;
uniform float scale_y;
out vec2 tex_coord;
void main () {
  gl_Position = vec4 (pos[0]*scale_x-1.0, pos[1]*scale_y-1.0, 0.0, 1.0);
  tex_coord = font_pos;
}

mrview: [DEBUG] compiling OpenGL fragment shader:
#version 330 core
in vec2 tex_coord;
uniform sampler2D sampler;
uniform float red, green, blue;
out vec4 color;
void main () {
  color.ra = texture (sampler, tex_coord).rg;
  color.rgb = color.r * vec3 (red, green, blue);
}

mrview: [DEBUG] font loaded
mrview: [INFO] opening image "image1.nii.gz"...
mrview: [DEBUG] sanitising image information...
mrview: [DEBUG] loading image "image1.nii.gz"...
mrview: [INFO] mrview: uncompressing image "image1.nii.gz"
mrview: [INFO] opening image "image2.nii.gz"...
mrview: [DEBUG] sanitising image information...
mrview: [DEBUG] loading image "image2.nii.gz"...
mrview: [INFO] mrview: uncompressing image "image2.nii.gz"
mrview: [INFO] mrview: uncompressing image "image2.nii.gz" [done]
mrview: [DEBUG] image "image2.nii.gz" loaded
mrview: [DEBUG] image "image2.nii.gz" initialised with strides = [ -1 98 9604 ], start = 97, using indirect IO
```

**Comment by jdtournier** [comment]
OK, I can reproduce the segfault using this command:

    mrview.exe in.nii.gz -overlay.load tmp.nii.gz

In this case, `in.nii.gz` is 4D, `tmp.nii.gz` is 3D. Debug output is essentially identical to @ansk's previous post. This is on a Windows 10 system - fully up to date (OS, MSYS2, and MRtrix3). 

I can also reproduce the interpolation option being ignored issue. 

What's happening seems to relate to whether the progress bar appears or not. If I reduce the size of `in.nii.gz` to say 12 volumes (`mrconvert in.nii.gz -coord 3 0:11 in2.nii.gz`), there is no issue, presumably because the decompression is fast enough that the progressbar doesn't get shown (it only shows after 1 sec). So this all points to issues with the OpenGL context getting switched between the main window and the progress dialog, so that subsequent OpenGL commands are directed at the wrong context. I'll have a look into it, hopefully it'll be relatively simple to fix. 

> I've also been unable to find the commit corresponding to your specific version; are you currently checked out on some other branch? Can you try reproducing the fault with an up-to-date installation?

Just for the record, that version dates back from Nov 23 (that extra `g` in front of the commit SHA1 in the version string is quite confusing, but it gets added by `git -describe` no matter what we do...): 

```
commit ff06655194c1590176269e8fd862f26d66ed3424
Merge: b7d9fad 2425b03
Author: Robert Smith <robert.smith@florey.edu.au>
Date:   Wed Nov 23 14:03:04 2016 +1100

    Merge pull request #831 from MRtrix3/mrtransform_directions

    Mrtransform directions
```

**Comment by jdtournier** [comment]
OK, this should hopefully be fixed with #875. The issue was indeed related to the progress dialog, but not quite in the way I'd anticipated. 

In case anyone's interested: the progress dialog needs to invoke `qApp->processEvents()` once per update to ensure the contents of the GUI are updated, but that also means any other pending events get processed too. The issue here was that the command-line parsing was performed as a one-shot timer, to be invoked at some point after window creation once the main event loop was up and running. But invoking `processEvents()` directly essentially runs one iteration of the event loop, so the net result is that the `overlay.load` option was being processed in the middle of the main image decompression (as soon as the corresponding progress dialog was shown)... 

I'll close this for now, please go to #875 for any follow-up discussion.
